### ISSUE TITLE
GET and DELETE request payloads sent but truncated when large

### ISSUE THREAD

**Issue by rogierc** [eclipse/californium#834]
When GET or DELETE requests contain payloads, the payloads are transferred to the server. When the payloads are large, there is - so it seems - no blockwise transfer done, causing truncation of the payload. No exception is thrown, which can be expected to cause problems on the user application level. 

Rfc7252 & rfc7959 don't define payloads in GET and DELETE requests. So, it's not clear to me how this is intended. Shurely, a truncated payload is not. Seems to me either:
- not transfering blockwise is not intended
- transfering payload alltogether is not intended and should throw an exception
- Californium user should not set payloads on these requests.

In the latter case (or latter 2 cases) Mule CoAP Connector - as Cf user - has to prohibit payloads in GET and DELETE requests.

What's your opinion on this?

Used Californium version: 2.0.0-M12

**Comment by boaks** [comment]
Californium offers two blockwise modes, transparent (done by californium) and "pass through" (blocks are passed through and the blocklogic is done in user-code on the top of californium). The modes are controlled by the NetworkConfig property "MAX_RESOURCE_BODY_SIZE". Please refer to the javadoc there (https://github.com/eclipse/californium/blob/2.0.x/californium-core/src/main/java/org/eclipse/californium/core/network/config/NetworkConfig.java#L171).

The usage blockwise may also be adjust/optimized , according your network constraints, with the PREFERRED_BLOCK_SIZE and MAX_MESSAGE_SIZE.

By the way, at least it's assumed, that: 
a GET request should not contain payload, the response will.
a DELETE request should not contain payload, the response also should not contain payload.
(a PUT request will contain payload, the response should not.
a POST request may contain payload, and the response may also contain payload).

**Comment by boaks** [comment]
> Rfc7252 & rfc7959 don't define payloads in GET and DELETE requests. So, it's not clear to me how this is intended. 

I guess your issue will be solved with setting the MAX_RESOURCE_BODY_SIZE to a proper value. Beside of "observes" (which are just bad idea's for blockwise transfers :-) ), the method is not considered for the blockwise. I also can't see that other code limits the use of payload related to the used method or return code. FMPOV it's at least good practice, to comply to the assumptions.

**Comment by rogierc** [comment]
> I guess your issue will be solved with setting the MAX_RESOURCE_BODY_SIZE

In my testcase MAX_RESOURCE_BODY_SIZE=8192  and payloads are this size or smaller. The GET/DELETE request payloads arrive truncated to 2012 bytes (also Fetch/Patch do this), the POST/PUT request payloads arrive, on the other hand, complete...

To clarify my case, I've made the test independent of the Mule Coap Connecter and put it on github:
https://github.com/rogierc/cf-issue-834/blob/master/src/test/java/nl/teslanet/test/cf/Issue834Test.java

In the test, a client does a GET, POST, PUT, DELETE, FETCH and PATCH request with small (10 bytes) and large (8192 bytes) payload. The server registers the length of received payload and returns this length in the response. The test reports:

Results :
Failed tests: 
* testGetWithLargePayload(nl.teslanet.test.cf.Issue834Test): wrong content length returned:  expected:<8192> but was:<2012>
* testDeleteWithLargePayload(nl.teslanet.test.cf.Issue834Test): wrong content length returned:  expected:<8192> but was:<2012>
* testPatchWithLargePayload(nl.teslanet.test.cf.Issue834Test): wrong content length returned:  expected:<8192> but was:<2012>
* testFetchWithLargePayload(nl.teslanet.test.cf.Issue834Test): wrong content length returned:  expected:<8192> but was:<2012>
Tests run: 12, Failures: 4, Errors: 0, Skipped: 0

> I also can't see that other code limits the use of payload related to the used method or return code. 

That would mean these testresults are unexpected, is it not?


> "observes" (which are just bad idea's for blockwise transfers :-) ),

Is this because of queueing of notifications that could occur on fast changing resources, or are there other reasons?

> FMPOV it's at least good practice, to comply to the assumptions.

I can only agree :-). What interests me is: why have Cf-developers chosen not to enforce the good practice and allow GET/DELETE requests to have a payload? This because I have to decide whether to allow or disallow it in the Mule Coap Connetor as well.

**Comment by boaks** [comment]
> Is this because of queueing of notifications that could occur on fast changing resources, or are there other reasons?

That queuing is always wrong for notifications according RFC7641. With blockwise this just get's much more unusable. Potentially, if 90% of the blocks are transferred, and the resource gets changed, all block are send in vain. It ends up very busy communication without a notify reaches the client. And then the people start to "make it working in their case".

**Comment by boaks** [comment]
> Failed tests:

strange, I will use your tests and check it.

**Comment by boaks** [comment]
I was wrong ;-(.

BlockwiseLayer:

```
	private boolean requiresBlockwise(final Request request) {
		boolean blockwiseRequired = false;
		if (request.getCode() == Code.PUT || request.getCode() == Code.POST) {
			blockwiseRequired = request.getPayloadSize() > maxMessageSize;
		}
		if (blockwiseRequired) {
			LOGGER.debug("request body [{}/{}] requires blockwise transfer", request.getPayloadSize(), maxMessageSize);
		}
		return blockwiseRequired;
	}
```

I searched for GET ... so I didn't find it.

**Comment by boaks** [comment]
> why have Cf-developers chosen not to enforce the good practice and allow GET/DELETE requests to have a payload? This because I have to decide whether to allow or disallow it in the Mule Coap Connector as well.

If it's enforced, you can't use it, even if the other peer requires it to do so.
May be we make it configurable.

**Comment by boaks** [comment]
Let's try to find out, what the other committers think about it. Unfortunately, that will take some time. Next week I will not be in the office.

**Comment by rogierc** [comment]
I 'm not in a hurry ;-)

Thanks!

**Comment by sbernard31** [comment]
I didn't find it explicitly in the [RFC 7252](https://tools.ietf.org/html/rfc7252) but It seems pretty clear that GET and DELETE is not done to send payload.

About block-wised transfer, the [RFC 7959 §2.1](https://tools.ietf.org/html/rfc7959#section-2.1) seems to confirm that : 
_"Hence, for the methods defined in [RFC7252], Block1 is useful with
the payload-bearing POST and PUT requests and their responses."_

I totally agree that the current behavior is not good and we should fix it.
My feeling : 
1. PATCH and FETCH should have payload and so I suppose it make sense to make it support transparent block-wised.
2. For DELETE and GET, I would be OK to not allow payload for it. But If there is use cases about using payload with this requests or if this makes the code easier we could allow it.
3. In all case, it would be good to understand what cause this "truncate" behavior ?

@boaks I didn't get what you want to make configurable ?

**Comment by boaks** [comment]
FMPOV, the current mismatch seems to be, that a GET and DELETE **CAN** have payload, but that is not supported to be blockwise.

If requests in general are strict about the "method/payload" rule, then it changes the current behavior. I'm not sure, if someone rely on such a "misuse". But if, then it's easier, if there is already a configurable switch to disable the strict check (again).

**Comment by sbernard31** [comment]
3. Ok so the "truncate" issue, is more about the way we handle large Datagram at UDPConnector side.
If packet receive size is > to UDP_CONNECTOR_DATAGRAM_SIZE, the behavior is not guaranty
I don't know if we can be better on this.

2. If we make "payload for GET/DELETE" configurable, do we make it works with blockwise ?

**Comment by boaks** [comment]
> 2- If we make "payload for GET/DELETE" configurable, do we make it works with blockwise ?

I would prefer to support it then also for blockwise. 
It would fix the mismatch in the current implementation. 

Just to mention it:
The Request itself has no access to the NetworkConfig. The errors could then only be reported, when the Request is "send" (CoapEndpoint).

Or the dirty trick:
"set the payload and after that report that Exception".
If someone catches that and send the Request anyway, then it would work even without configuration (but some extra work, for those who use GET/DELETE with payload).

**Comment by boaks** [comment]
Or add a "force payload" flag to the Request.
The someone would need to set that before setting payload for GET/DELETE in order to be not faced the "strict payload rules denied" exception.

**Comment by sbernard31** [comment]
> I would prefer to support it then also for blockwise.
Same for me.

> Or add a "force payload" flag to the Request.
The someone would need to set that before setting payload for GET/DELETE in order to be not faced the "strict payload rules denied" exception.

I like the idea.

**Comment by boaks** [comment]
> Or add a "force payload" flag to the Request.
> Then someone would need to set that before setting payload for GET/DELETE in order to be not faced the "strict payload rules denied" exception.
> 

> I like the idea.

I started with that today.

**Comment by boaks** [comment]
Using PR #844 and adapting the tests above with:

```
	public void runTestCase(Code code, byte[] payload, ResponseCode expect) {
		Request request = new Request(code);
		if (!request.isIntendedPayload()) {
			request.setUnintendedPayload();
		}
		request.setPayload(payload);
```

should show the new intended behavior.

**Comment by sbernard31** [comment]
About the silent truncation, could we enlarge the DatagramBuffer to UDP_CONNECTOR_DATAGRAM_SIZE + 1 ?

This way we could check that if datagram received is larger than the expected size (here UDP_CONNECTOR_DATAGRAM_SIZE).

If the datagram is too large, we drop it and log a warning ?

**Comment by boaks** [comment]
Currently I'm not really concerned about too large messages.
But, yes, if you want, just create a PR.

**Comment by boaks** [comment]
@rogierc 

Do you have a ECA? If so, could you provide your unit tests as PR?

**Comment by rogierc** [comment]
I don't have an ECA... Is it an option to use the unittest by copying the code?

**Comment by rogierc** [comment]
Or is it difficult to get an ECA?

**Comment by rogierc** [comment]
It'not difficult ;-) I have one now. Will make a PR, is there anything else I need to do that?

**Comment by boaks** [comment]
"Sign" the commit and PR. It's simply:

```
Signed-off-by: firstname lastname email@address
```

**Comment by boaks** [comment]
Fixed with PR #847, tests in PR #849.
